src/Playwright/Transport/WebSocketTransport.cs

```diff
@@ -0,0 +1,199 @@
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Close(ex);
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Close(string closeReason)
+        {
+            if (!IsClosed)
+            {
+                IsClosed = true;
+                _ = _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, closeReason, _readerCancellationSource.Token).ConfigureAwait(false);
+                TransportClosed?.Invoke(this, new TransportClosedEventArgs { CloseReason = closeReason });
+                _readerCancellationSource?.Cancel();
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Close(Exception ex)
+        {
+            System.Diagnostics.Debug.WriteLine(ex);
+            Close(ex.ToString());
+        }
+
+        private async Task ConnectAsync()
+        {
+            await _webSocket.ConnectAsync(new Uri(_wsEndpoint), _readerCancellationSource.Token).ConfigureAwait(false);
+            ScheduleTransportTask(ReceiveAsync, _readerCancellationSource.Token);
+        }
+
+        private void ScheduleTransportTask(Func<CancellationToken, Task> func, CancellationToken cancellationToken)
+            => Task.Factory.StartNew(() => func(cancellationToken), cancellationToken, TaskCreationOptions.LongRunning, TaskScheduler.Current);
+
+        private async Task ReceiveAsync(CancellationToken token)
+        {
+            try
+            {
+                if (_webSocket.State == WebSocketState.Closed)
+                {
+                    Close("Closed");
+                }
+                else
+                {
+                    var buffer = WebSocket.CreateClientBuffer(DefaultBufferSize, DefaultBufferSize);
+
+                    while (!token.IsCancellationRequested && _webSocket.State == WebSocketState.Open)
+                    {
+                        WebSocketReceiveResult result;
+                        using (MemoryStream memoryStream = new())
+                        {
+                            do
+                            {
+                                result = await _webSocket.ReceiveAsync(buffer, token).ConfigureAwait(false);
+
+                                if (result.MessageType == WebSocketMessageType.Close)
+                                {
+                                    await
+                                        _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, string.Empty, _readerCancellationSource.Token).ConfigureAwait(false);
+                                }
+                                else
+                                {
+#if NETSTANDARD
+                                    await memoryStream.WriteAsync(buffer.Array, 0, result.Count, _readerCancellationSource.Token).ConfigureAwait(false);
```

Comment:
I don't think memcpy should be async.

Reply:
This will be async since we are calling from the async method. `Write` synchronously blocks hence have to use async here.

Reply:
What is wrong with blocking? We are presumably copying a chunk of memory. Maybe the MemoryStream is not the best data structure for it?

Reply:
It looks like MemoryStream is ok, just don't use the async versions of these methods.

Reply:
Updated to synchronous call

Reply:
fixed